File: frontend/demo/component/passwordfield/password-field-helper.ts
import 'Frontend/demo/init'; // hidden-source-line

import { html, LitElement } from 'lit';
import { customElement } from 'lit/decorators.js';
import '@vaadin/vaadin-text-field/vaadin-password-field';
import { applyTheme } from 'Frontend/generated/theme';

@customElement('input-field-helper')
export class Example extends LitElement {
  protected createRenderRoot() {
    const root = super.createRenderRoot();
    // Apply custom theme (only supported if your app uses one)
    applyTheme(root);
    return root;
  }

  render() {
    return html`
      <!-- tag::snippet[] -->
      <vaadin-password-field
        label="Password"
        helper-text="A password must be at least 8 characters.
          It has to have at least one letter and one digit."
        pattern="^(?=.*[0-9])(?=.*[a-zA-Z]).{8}.*$"
        error-message="Not a valid password"
      >
      </vaadin-password-field>
      <!-- end::snippet[] -->
    `;
  }
}

File: src/main/java/com/vaadin/demo/component/passwordfield/PasswordFieldHelper.java
package com.vaadin.demo.component.passwordfield;

import com.vaadin.flow.component.html.Div;
import com.vaadin.flow.component.textfield.PasswordField;
import com.vaadin.flow.router.Route;
import com.vaadin.demo.DemoExporter; // hidden-source-line

@Route("password-field-helper")
public class PasswordFieldHelper extends Div {

  public PasswordFieldHelper() {
    // tag::snippet[]
    PasswordField passwordField = new PasswordField();
    passwordField.setLabel("Password");
    passwordField.setHelperText("A password must be at least 8 characters. It has to have at least one letter and one digit.");
    passwordField.setPattern("^(?=.*[0-9])(?=.*[a-zA-Z]).{8}.*$");
    passwordField.setErrorMessage("Not a valid password");
    add(passwordField);
    // end::snippet[]
  }

  public static class Exporter extends DemoExporter<PasswordFieldHelper> { // hidden-source-line
  } // hidden-source-line
}
